### ISSUE TITLE
Navigator forces expand size constraint.

### ISSUE THREAD

**Issue by slightfoot** [flutter/flutter#14803]
I found that I wanted to use the a Step like UX where the user can proceed through a series of steps. The back button on Android should take them to the previous step. This should appear in a dialog.

I found when you use a `Navigator` it forces the contents to max constraints, which means the dialog must fill the screen. I feel that Navigator is a almost has a "Semantic" meaning that it controls the routes and history back stack, and its parent should constrain where the stack/transitions occur. I believe this is a bug.

Some digging I found that `Navigator` returns an `Overlay` widget that contains a `Stack` with `fit: StackFit.expand`. From what I can from the Flutter Inspector this is the root cause.

I've included a small example below for testing. I'm not setup to build the engine repo, or I would have tested my supposition above. It would be great to get some feedback on if this is the problem and if an optional fit parameter put on Overlay so it could be optional.

@Hixie I assume you want to know about this for your Navigator work.

## Steps to Reproduce
**Example Code**
```dart
import 'package:flutter/material.dart';


void main() => runApp(new MainWidget());

class MainWidget extends StatelessWidget {
	@override
	Widget build(BuildContext context) {
		return new MaterialApp(
			home: new Scaffold(
				backgroundColor: Colors.lightBlue,
				body: new AlertDialog(
					title: const Text('Help'),
					content: new Navigator(
						initialRoute: 'step:1',
						onGenerateRoute: _onGenerateRoute,
					),
				),
			),
		);
	}

	Route<dynamic> _onGenerateRoute(RouteSettings settings) {
		var parts = settings.name.split(':');
		return new PageRouteBuilder(
			pageBuilder: (_, __, ___) => new StepWidget(int.parse(parts[1], radix: 10)),
			transitionsBuilder: (BuildContext _, Animation<double> animation, Animation<double> secondary, Widget child) {
				child = new SlideTransition(
					position: new Tween<Offset>(
						begin: Offset.zero,
						end: const Offset(-1.0, 0.0),
					).animate(secondary),
					child: child,
				);
				child = new SlideTransition(
					position: new Tween<Offset>(
						begin: const Offset(1.0, 0.0),
						end: Offset.zero,
					).animate(animation),
					child: child,
				);
				return child;
			}
		);
	}
}

class StepWidget extends StatelessWidget {
	final int step;

	StepWidget(this.step);

	@override
	Widget build(BuildContext context) {
		return new Column(
			mainAxisSize: MainAxisSize.min,
			children: <Widget>[
				new Container(
					padding: const EdgeInsets.all(48.0),
					child: new Text('Step $step'),
				),
				new Row(
					mainAxisSize: MainAxisSize.max,
					children: <Widget>[
						new Expanded(
							child: new FlatButton(
								child: const Text('Prev'),
								onPressed: () => Navigator.of(context).maybePop(),
							),
						),
						new Expanded(
							child: new FlatButton(
								child: const Text('Next'),
								onPressed: () => Navigator.of(context).pushNamed('step:${step + 1}'),
							),
						),
					],
				),
			],
		);
	}
}
```

## Flutter Doctor
```
√] Flutter (on Microsoft Windows [Version 10.0.14393], locale en-GB, channel master)
    • Flutter version 0.1.1-pre.33 at C:\Android\flutter
    • Framework revision d401bd78ae (8 days ago), 2018-02-13 01:56:13 -0800
    • Engine revision f5a4a93787
    • Tools Dart version 2.0.0-dev.22.0
    • Engine Dart version 2.0.0-edge.32396138ff733cbe8d9ac32178e6cffccdf80644

[√] Android toolchain - develop for Android devices (Android SDK 27.0.3)
    • Android SDK at C:\Android\SDK
    • Android NDK at C:\Android\SDK\ndk-bundle
    • Platform android-27, build-tools 27.0.3
    • ANDROID_HOME = C:\Android\SDK
    • Java binary at: C:\Android\Studio\jre\bin\java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b01)

[√] Android Studio (version 3.0)
    • Android Studio at C:\Android\Studio
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b01)

[√] IntelliJ IDEA Community Edition (version 2017.3)
    • Flutter plugin version 21.2.3
    • Dart plugin version 173.4548.30

[√] Connected devices
    • Google Nexus 5X, 7 1 0, API 25, 1080x1920 • 192.168.188.101:5555 • android-x86 • Android 7.1.1 (API 25)

• No issues found!
```

**Comment by yacineblr** [comment]
Hi, Is it possible to make the browser adapt to the size of the child?

**Comment by lulupointu** [comment]
Hi,

<details>
<summary>
Here is an even simpler example to showcase this issue
</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      color: Colors.white,
      home: Center(
        child: Navigator(
          pages: [
            MaterialPage(
              child: ProfileWidget(),
            ),
          ],
          onPopPage: (_, __) {
            return false;
          },
        ),
      ),
    );
  }
}

class ProfileWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Container(
      color: Colors.red,
      height: 200,
    );
  }
}
```
</details>


<details>
<summary>
Given result
</summary>

![image](https://user-images.githubusercontent.com/5128921/116660121-dfccea80-a992-11eb-8be2-9a6f230be5cf.png)

</details>

<details>
<summary>
Expected result
</summary>

![image](https://user-images.githubusercontent.com/5128921/116660053-c166ef00-a992-11eb-81e2-c12a930497e9.png)

</details>

The **Expected result** was obtained be moving the `Center` widget in the `MaterialPage`.

This is problematic because it means that we would have to duplicate this constraint in every `Page` rather than having it only once on top of `Navigator`.

Could we at least have some feedback on whether this is considered as a bug and will be fixed or if we should just duplicate and move the constraint down?

**Comment by CosmicPangolin** [comment]
@Hixie 

This might have higher priority now that people are trying to build wrappers around navigator 2.0

**Comment by T-Oner** [comment]
Is there any progress on this issue？

**Comment by raulmabe** [comment]
> Is there any progress on this issue？

Doesn't seem so... I am also interested on this topic. Does anyone know a workaround?

**Comment by cirnok** [comment]
Because of this problem, it is impossible to implement nested navigation in dialogs without expanding them to full screen. If this UI is a required part of the application, developers will not be able to implement it and will have to change Flutter to something else.

**Comment by Hixie** [comment]
I'm not sure I really understand the issue here. You can wrap the Navigator in whatever widget you like to change the constraints passed to the Navigator, and you can wrap the Navigator's children with whatever widgets you like to have them treat their constraints however is desired, including using `Positioned` to make them not be expanded by default. If someone wanted to make the Overlay and Navigator widgets configurable so that e.g. Overlay's `computeDryLayout` didn't just always use `constraints.biggest`, that might also be possible (it would depend on the benchmark results), but it would be good to better explain the use case first.

**Comment by DevNico** [comment]
We've just stumbled upon this issue trying to implement navigation in a bottom sheet which should in theory automatically adjust its height to the current route. Could you provide an example of how to solve this? @Hixie

This stackoverflow question has some good imagery to explain the issue: https://stackoverflow.com/questions/68964864/make-nested-navigator-widget-wrap-child-in-column

**Comment by yellowgraysu** [comment]
I recently found a workaround but not sure if it is good to use.

Just wrap the Navigator with [IntrinsicHeight](https://api.flutter.dev/flutter/widgets/IntrinsicHeight-class.html) (which was warned that the class is relatively expensive).

**Comment by saibotma** [comment]
Yeah, this is a workaround. Unfortunately, this does not work when you use `LayoutBuilder` below `Navigator`, because `LayoutBuilder` can not calculate intrinsic sizes.

**Comment by paulkre** [comment]
Does anyone know a package that provides this feature? It's very sad that there is currently no way to do this in Flutter. I really don't care whether the implementation is efficient or not.

**Comment by nblum37** [comment]
Hi all,
same issue here. I would really like to avoid using `IntrinsicHeight`.

**Comment by adonisRodxander** [comment]
Same issue

**Comment by lukaskurz** [comment]
Same here, i took me considerable time to figure out that the issue was not related to auto_route, but with flutter itself.

**Comment by nblum37** [comment]
Is there any possibility to push this issue? :)

**Comment by Hixie** [comment]
@nblum37 The primary way to push this issue would be to contract with someone to have them do the work (or to contribute a patch directly). See also https://github.com/flutter/flutter/wiki/Issue-hygiene#when-will-my-bug-be-fixed

**Comment by saibotma** [comment]
Just stumbled upon this issue again while building another app. 
If it is just a matter of removing `StackFit.expand` on the `Overlay` inside `Navigator` then this should be an easy fix, or is there more that needs to be done to not cause a regression?
